### PR TITLE
Introduce a configuration option to define the internaldate used for APPEND, Fixes #522

### LINKED ISSUE #522: Offlineimap sets wrong internaldate
Hello,
I'm using Offlineimap, my mail provider is [FastMail](https://www.fastmail.com/). Whenever I'm moving a file around locally (from one Maildir directory into another one) the mail is first being deleted on the server, and then uploaded again, so far so good. But when the mail is uploaded again the internaldate of the message get's changed to the current timestamp:

`2018-02-06 13:28:26 DEBUG: [imap]: savemessage: date: "06-Feb-2018 13:27:33 +0100", […]`

So in the end the mail shows on FastMail with the wrong date. Correct woud be the date which is set in the Date header of the mail. How can I configure Offlineimap to respect the Date header here?

### PR DESCRIPTION
**Pull request by jgonsior**
### This PR

- [x] I've read the [DCO](http://www.offlineimap.org/doc/dco.html).
- [x] I've read the [Coding Guidelines](http://www.offlineimap.org/doc/CodingGuidelines.html)
- [ ] The relevant informations about the changes stands in the commit message, not here in the message of the pull request.
- [ ] Code changes follow the style of the files they change.
- [ ] Code is tested (provide details).

### References

- Issue #522

### CODE REVIEW

**Comment by nicolas33** [comment]
There should be no merge in your request and no duplicated patches. Please, rebase your work with `git rebase` on top of the current "next" or "master" branch.

About the patch, be aware that the rtime value is critical code for more than one option AFAIR. This must be handled with care.

Currently, it looks like the rtime parameter is now ignored. Is this correct?

I think we should explain the new fallback strategy in the offlineimap.conf file.

Once the patch is ready don't forget to sign your work. ,-)

File: offlineimap.conf [hunk 1]
```diff
@@ -1066,7 +1066,11 @@ remoteuser = username
 #   - header (the Date header of the email)
 #   - imap (let the IMAP server set the internaldate)
 #
-# Default is timestamp
+# If nothing is set, or the given option is failing for some reasens, the following
```
Comment (nicolas33) [inline offlineimap.conf]: reasons

Comment (jgonsior) [reply offlineimap.conf]: I thought that was the way offlineimap behaved before?

File: offlineimap/folder/IMAP.py [hunk 2]
```diff
@@ -534,11 +533,17 @@ def __getmessageinternaldate(self, content, rtime=None):
...
+        # if config parameter is set respect this setting:
+        # for the default value we don't need to change anything
+        internal_date_for_append_method = self.repository.getconf('internal_date_for_append_method', 'timestamp')
 
-        if rtime is None:
+        if rtime is None or internal_date_for_append_method == 'header':
             rtime = emailutil.get_message_date(content)
-            if rtime == None:
+        
+        if internal_date_for_append_method == 'imap' or rtime == None:
```
Comment (nicolas33) [inline offlineimap/folder/IMAP.py]: `rtime is None`

File: offlineimap.conf (hunk 1 above)
Comment (nicolas33) [reply offlineimap.conf]: It's about the syntax. What's written is "reasens".

**Comment by nicolas33** [comment]
I'm not sure about this version. Could you explain why the use cases where rtime must or must not be considered to compute the returned value?

**Comment by nicolas33** [comment]
About rtime, please refer to the comments in the code. I'd like to know when it's correct to ignore rtime since rtime is already set for some use cases (likely to honor configuration options). I wonder there should be comments in the code about that.

**Comment by jgonsior** [comment]
Hm, no idea actually. I tried to change the function in a way that, if the newly added internal_date_for_append_method configuration option is set, we do what the option says, and if it is not set I tried to keep the original code as it is.
If there is any case where the newly added option shouldn't be applied but the given parameter instead - I don't know.

**Comment by nicolas33** [comment]
So, we have to check if there are conflicts and how to solve them, if any.

**Comment by nicolas33** [comment]
rtime is the variable to define the datetime. For imap, we should rather define the fallback strategy between:

- timestamp, header, imap
- header, timestamp, imap
- header, imap
- timestamp, imap

Hence, I think the configuration option should be `imap_datetime_strategy = A, B` where `A, B` is a list (possibly `None`). Internally, we would always add the imap strategy in the end.

In order to be merged, the PR must be correctly written (rebased in this case) with the DCO line(s) set.
That's why I'm closing this PR.

**Comment by schiebel** [comment]
Thanks @jgonsior for these changes. I found your fork useful!

**Pull request closed without merging**